### PR TITLE
304 response should not include Content-Type header

### PR DESCRIPTION
**Pull request by eagletmt**
Rack::Lint raises an error saying "Content-Type header found in 304
response, not allowed".

### CODE REVIEW

File: actionpack/lib/action_dispatch/middleware/static.rb [hunk 1]
```diff
@@ -48,7 +48,10 @@ def call(env)
         env['PATH_INFO']            = gzip_path
         status, headers, body       = @file_server.call(env)
         headers['Content-Encoding'] = 'gzip'
-        headers['Content-Type']     = content_type(path)
+        if status != 304
+          # 304 response should not include Content-Type header.
```
Comment (arthurnn) [inline actionpack/lib/action_dispatch/middleware/static.rb]: I dont think the comment is necessary here, as the `if` already states that statement .

**Comment by arthurnn** [comment]
As the response status code is 304, I wonder if we could short-cut this middleware altogether. 
cc @jeremy

File: actionpack/lib/action_dispatch/middleware/static.rb (hunk 1 above)
Comment (eagletmt) [inline actionpack/lib/action_dispatch/middleware/static.rb]: OK, I removed the trivial comment and force-pushed.

**Comment by rafaelfranca** [comment]
Yeah, I think we should skip most of the work of this middleware if 304

**Comment by pixeltrix** [comment]
@rafaelfranca yes, since `Rack::File` just returns [an empty triplet][1]. @eagletmt can you rewrite `call` so that all  of the response headers are only added if it's not 304, thanks.

Of course it'd be much better if this lived in `Rack::File` itself where it should be :cry:

[1]: https://github.com/rack/rack/blob/master/lib/rack/file.rb#L64

**Comment by wyaeld** [comment]
When did this behaviour in either rack or rails change?

**Comment by eagletmt** [comment]
@arthurnn @rafaelfranca @pixeltrix I'm sorry but I couldn't figure out what to do.
Content-Encoding header must be set because FileHandler returns the content of `/path/to/foo.js.gz` in response to `/path/to/foo.js` .
What kind of procedure can I skip or short-cut here?

**Comment by arthurnn** [comment]
As far as I know, if the status code is 304, you wont need to set any header nor a body in the response. So you could return it early.

**Comment by eagletmt** [comment]
Ah, I get it. I modified to skip Content-Encoding and Vary header.
Thank you for your advice.

**Comment by arthurnn** [comment]
thanks @eagletmt

**Comment by adrianpike** [comment]
@wyaeld I actually just ran into this bug, and was about to open a ticket. #16466 is where it started, looks like it went out in 4.2.

Looks like I get to live against master for a little bit. :)

**Comment by wyaeld** [comment]
@adrianpike thanks for the confirmation.  We took the lesser evil temporarily of just including `middleware/static.rb` from master in the app, which we'll rip out once its fixed.  That works fine for now.

**Comment by arthurnn** [comment]
Added a changelog entry for this https://github.com/rails/rails/commit/b6b088429c945b5c0c213fcdb9b4f6658faf686a .
And I backport it to 4-2-stable : https://github.com/rails/rails/commit/061b487b460e091fb99bf35cd147eeb7b5c02e23

**Pull request merged**